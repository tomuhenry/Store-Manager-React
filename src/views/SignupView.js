import React, { Component } from 'react';
import SignupForm from '../components/SignupForm';
import { connect } from 'react-redux';
import signupAction from '../actions/signupAction';
import Header from '../components/Header';
import { toast } from 'react-toastify';
import Loader from '../components/Loader';

export class SignupView extends Component {
  state = {
    value: '',
    errors: {},
    data: '',
    loading: false
  };

  componentDidMount() {
    const token = localStorage.getItem('access_token')
    if (!token) {
      toast.error('Please login');
      this.props.history.push('/');
    }
  }

  handleChange = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  componentWillReceiveProps(nextProps) {
    const { data, errors } = nextProps;
    if (errors) {
      this.setState({ errors: errors });
      this.setState({ loading: false });
    } else if (data) {
      this.setState({ data:data });
      this.setState({ loading: false });
    }
  }

  handleFormSubmit = e => {
    e.preventDefault();
    const data = {
      email: this.state.email,
      name: this.state.name,
      password: this.state.password
    };
    this.setState({ loading: true });
    this.props.signupAction(data);
  };

  render() {
    return (
      <div>
        <Header />
        <p id='welcome_text'>Add new user to Store Manager</p>
        <div className='loader'>{this.state.loading ? <Loader /> : null}</div>
        <SignupForm
          changed={this.handleChange}
          FormSubmit={this.handleFormSubmit}
          errors={this.state.errors}
          data={this.state.data}
        />
      </div>
    );
  }
}
export const mapStateToProps = state => {
  return {
    data: state.signup.data,
    errors: state.signup.errors
  };
};
export default connect(
  mapStateToProps,
  { signupAction }
)(SignupView);
