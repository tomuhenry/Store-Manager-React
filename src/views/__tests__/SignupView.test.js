import React from 'react';
import { shallow } from 'enzyme';
import {SignupView} from '../SignupView';

describe('Signup View', () => {
  let wrapper;
  const mockSignUpfn = jest.fn();
  const historyMock = { push: jest.fn() };

  beforeEach(() => {
    wrapper = shallow(
      <SignupView signupAction={mockSignUpfn} history={historyMock} />
    );
  });
  afterEach(() => {
    mockSignUpfn.mock.calls = [];
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.length).toEqual(1);
  });

  it('should be call fetchLogin with the username and password in the state as arguments', () => {

    wrapper.find('SignupForm').dive().find('#email').simulate('change', {
      target: { "email": "value@email.com" }
    });
    wrapper.find('SignupForm').dive().find('#name').simulate('change', {
      target: { "name": "value" }
    });
    wrapper.find('SignupForm').dive().find('#password').simulate('change', {
      target: { "password": "value" }
    });
    wrapper.find('SignupForm').dive().find('form').simulate('submit', { preventDefault() {} });
    expect(mockSignUpfn.mock.calls.length).toEqual(1);

  });
  it('Should recieve props data', () => {
    const spy = jest.spyOn(SignupView.prototype, 'componentWillReceiveProps');
    const nextProps={
      data: {data: true}
    }
    wrapper.instance().componentWillReceiveProps(nextProps);
    wrapper.setProps({
      loading: true
    })
    expect(spy).toHaveBeenCalled()
  })
  it('Should recieve props error', () => {
    const spy = jest.spyOn(SignupView.prototype, 'componentWillReceiveProps');
    const nextProps={
      errors: {errors: true}
    }
    wrapper.instance().componentWillReceiveProps(nextProps);
    wrapper.setProps({
      loading: true,
      errors: true
    })
    expect(spy).toHaveBeenCalled()
  })

});
