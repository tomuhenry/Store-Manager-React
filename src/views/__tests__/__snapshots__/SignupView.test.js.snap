// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup View should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
